### PR TITLE
feat(argocd-apps)!: use maps instead of lists

### PR DESCRIPTION
**Pull request by irizzant**
This PR fixes https://github.com/argoproj/argo-helm/issues/2523 

Checklist:

* [X] I have bumped the chart version according to [versioning](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#versioning)
* [X] I have updated the documentation according to [documentation](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#documentation)
* [X] I have updated the chart changelog with all the changes that come with this pull request according to [changelog](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#changelog).
* [X] Any new values are backwards compatible and/or have sensible default.
* [X] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md).
* [X] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/stable/developer-guide/ci/)).

### CODE REVIEW

File: charts/argocd-apps/Chart.yaml [hunk 1]
```diff
@@ -17,5 +17,12 @@ annotations:
     fingerprint: 2B8F22F57260EFA67BE1C5824B11F800CD9D2252
     url: https://argoproj.github.io/argo-helm/pgp_keys.asc
   artifacthub.io/changes: |
+    - kind: added
+      description: make the chart use maps instead of lists
+      links:
+        - name: GitHub Issue
+          url: https://github.com/argoproj/argo-helm/issues/2523
+        - name: GitHub PR
+          url: https://github.com/argoproj/argo-helm/pull/2538
```
Comment (mkilchhofer) [inline charts/argocd-apps/Chart.yaml:26]: Can you remove the links, we do not use them :-)

Also please drop the lines 27 and 28 (only the annotations from the current PRs needs to stay here)

**Review by mkilchhofer — CHANGES_REQUESTED** [review]
Almost LGTM.

Are we confident that this solves all our issues? ;) @yu-croco @jmeridth are you also okay with this change?

File: charts/argocd-apps/Chart.yaml (hunk 1 above)
Comment (irizzant) [reply charts/argocd-apps/Chart.yaml]: Hi, done as requested

Comment (mkilchhofer) [reply charts/argocd-apps/Chart.yaml]: Lines below your change (L22/23) still needs to be removed

**Comment by jmeridth** [comment]
@mkilchhofer Looks good but wondering one thing:

`range $k, $v := .Values.projects` is in many place now but the `$v` (value) is not used in most of them.  Should we prefer `range $k, _ := .Values.projects` instead or am I missing it?

File: charts/argocd-apps/Chart.yaml (hunk 1 above)
Comment (irizzant) [reply charts/argocd-apps/Chart.yaml]: Hi @mkilchhofer done as requested

**Comment by irizzant** [comment]
@jmeridth I updated the manifests to use `range $k, _ := `

File: charts/argocd-apps/templates/item-templates.yaml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-{{- range .Values.itemTemplates }}
+{{- range $k, _:= .Values.itemTemplates }}
```
Comment (jmeridth) [inline charts/argocd-apps/templates/item-templates.yaml:1]: are you using `$k` anywhere in the file now?

**Review by jmeridth — COMMENTED** [review]
couple questions

File: charts/argocd-apps/Chart.yaml [hunk 3]
```diff
@@ -18,4 +18,4 @@ annotations:
     url: https://argoproj.github.io/argo-helm/pgp_keys.asc
   artifacthub.io/changes: |
     - kind: added
-      description: Support Template Patch to ApplicationSet
+      description: make the chart use maps instead of lists
```
[the remark this answers is not in the document]
Comment (irizzant) [reply charts/argocd-apps/Chart.yaml:21]: Done

File: charts/argocd-apps/templates/applications.yaml [hunk 4]
```diff
@@ -1,9 +1,9 @@
-{{- range .Values.applications }}
+{{- range $k, _:= .Values.applications }}
 ---
 apiVersion: argoproj.io/v1alpha1
 kind: Application
 metadata:
-  {{- with .additionalAnnotations }}
+  {{- with $v.additionalAnnotations }}
```
[the remark this answers is not in the document]
Comment (irizzant) [reply charts/argocd-apps/templates/applications.yaml:6]: done

File: charts/argocd-apps/templates/item-templates.yaml (hunk 2 above)
Comment (irizzant) [reply charts/argocd-apps/templates/item-templates.yaml:1]: Done

File: charts/argocd-apps/values.yaml [hunk 5]
```diff
@@ -1,253 +1,256 @@
...
+#  example:
+#    namespace: argocd
+#    additionalLabels: {}
+#    additionalAnnotations: {}
+#    sources:
+#      - git:
+#          url: https://github.com/argoproj-labs/argocd-example-extension.git
+#      - web:
+#          url: https://github.com/argoproj-labs/argocd-example-extension/releases/download/v0.1.0/extension.tar
+#
```
Comment (mkilchhofer) [inline charts/argocd-apps/values.yaml:256]: Please fix the chart-test error due to wrong EOF

Comment (irizzant) [reply charts/argocd-apps/values.yaml]: @mkilchhofer done as requested

**Comment by irizzant** [comment]
Any news ?

**Comment by jmeridth** [comment]
@mkilchhofer if this passes I'm good with it.

**Comment by irizzant** [comment]
@jmeridth is this on my end?

**Comment by jmeridth** [comment]
@irizzant yes, something to do [here](https://github.com/argoproj/argo-helm/pull/2538/files#diff-f7549c590d242f2340ae6c8ac25569a3c8196382f17a09149c6b28cb569c2c78R1).  I'm afk currently, so can't look.

**Review by mkilchhofer — CHANGES_REQUESTED** [review]
You didnt use the range function properly on all implemented resources.

```
$ helm template apps .
Error: parse error at (argocd-apps/templates/projects.yaml:1): range can only initialize variables
```

I fixed the project.yaml and run template again, then:
```
$ helm template apps .
Error: parse error at (argocd-apps/templates/item-templates.yaml:1): function "_" not defined

Use --debug flag to render out invalid YAML
```

I try to fix them

**Comment by mkilchhofer** [comment]
I reverted the changes on the `itemTemplates` section which was contributed by @joelee2012 in PR:
- #2025

Can you live with this @irizzant ? IMHO it is too complex to refactor this feature.

**Review by mkilchhofer — APPROVED** [review]

**Review by yu-croco — APPROVED** [review]
Thank you! 🚀

**Comment by irizzant** [comment]
@mkilchhofer thank you for the refactoring.
The need for this PR comes for being able to override ArgoCD resources with helm value files, I'm not against leaving the `itemTemplates` out of this PR also because I see it uses nested lists.

**Review by mbevc1 — APPROVED** [review]

**Comment by sathieu** [comment]
@irizzant Probably a good opportunity to add [schema](https://helm.sh/docs/topics/charts/#schema-files)?

**Pull request merged**